How do you use the hashsum_download api?

I'm trying to use girder as a midas replacement/cmake externalData module storage. @mgrauer mentioned this could be done using the `hashsum_download` plugin, but there doesn't seem to be any documentation for it. 

Could you guys share an example curl command or something showing how to use it?

So I get tripped up at the `Rebuild web code` step, which fails with an error (Due to proxy maybe?). When I try it on the command line, I get this:

```
GIRDER_LOCAL_FONTELLO_ARCHIVE=/root/girder/clients/web/static/built/fontello/fontello.zip girder-install web                           Connected to MongoDB: mongodb://localhost:27017/girder
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.0.0 (node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.0.14: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN grunt-webpack@1.0.18 requires a peer of webpack@1.x but none was installed.
npm WARN grunt-webpack@1.0.18 requires a peer of webpack-dev-server@1.x but none was installed.

> girder@2.0.0 build /root/girder_env/lib/python2.7/site-packages/girder
> grunt init && grunt "--env=prod" "--plugins=hashsum_download"

Using local fontello archive: /root/girder/clients/web/static/built/fontello/fontello.zip

Running "setup" task

Running "copy:fontello" (copy) task
Copied 1 files

Running "unzip:fontello" (unzip) task
Created "clients/web/static/built/fontello/" directory

Running "copy:swagger" (copy) task
Created 3 directories, copied 28 files

Done, without errors.
Using local fontello archive: /root/girder/clients/web/static/built/fontello/fontello.zip
Configuring plugin hashsum_download (plugin.json)

Running "plugin:hashsum_download" (plugin) task

Running "copy:plugin-hashsum_download" (copy) task


Running "stylus:swagger" (stylus) task
>> 1 file created.

Running "gitinfo" task

Running "file-creator:python-version" (file-creator) task
Writing file [girder/girder-version.json]
>> 1 file(s) written.

Running "file-creator:javascript-version" (file-creator) task
Writing file [clients/web/src/version.js]
>> 1 file(s) written.

Running "webpack:prod" (webpack) task
Version: webpack 2.1.0-beta.21         
                                             Asset       Size  Chunks             Chunk Names
                                 girder.app.min.js     913 kB       0  [emitted]  girder.app
  assets/glyphicons-halflings-regular-f4769f9b.eot    20.1 kB          [emitted]  
assets/glyphicons-halflings-regular-448c34a5.woff2      18 kB          [emitted]  
  assets/glyphicons-halflings-regular-e18bbf61.ttf    45.4 kB          [emitted]  
  assets/glyphicons-halflings-regular-89889688.svg     109 kB          [emitted]  
              assets/jsoneditor-icons-bf23347e.png    14.4 kB          [emitted]  
 assets/glyphicons-halflings-regular-fa277232.woff    23.4 kB          [emitted]  
            plugins/hashsum_download/plugin.min.js    16.6 kB       1  [emitted]  plugins/hashsum_download/plugin
                                 girder.ext.min.js    1.23 MB       2  [emitted]  girder.ext
                                girder.app.min.css    34.8 kB       0  [emitted]  girder.app
                                girder.ext.min.css     144 kB       2  [emitted]  girder.ext
           plugins/hashsum_download/plugin.min.css  128 bytes       1  [emitted]  plugins/hashsum_download/plugin

ERROR in girder.app.min.js from UglifyJs
SyntaxError: Unexpected token: punc ()) [girder.app.min.js:7866,21]
Warning: Task "webpack:prod" failed. Use --force to continue.

Aborted due to warnings.

npm ERR! Linux 4.2.0-42-generic
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "run" "build" "--" "--env=prod" "--plugins=hashsum_download"
npm ERR! node v6.8.1
npm ERR! npm  v3.10.9
npm ERR! code ELIFECYCLE
npm ERR! girder@2.0.0 build: `grunt init && grunt "--env=prod" "--plugins=hashsum_download"`
npm ERR! Exit status 3
npm ERR! 
npm ERR! Failed at the girder@2.0.0 build script 'grunt init && grunt "--env=prod" "--plugins=hashsum_download"'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the girder package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     grunt init && grunt "--env=prod" "--plugins=hashsum_download"
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs girder
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls girder
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /root/girder_env/lib/python2.7/site-packages/girder/npm-debug.log
Traceback (most recent call last):
  File "/root/girder_env/bin/girder-install", line 11, in <module>
    load_entry_point('girder', 'console_scripts', 'girder-install')()
  File "/root/girder_env/local/lib/python2.7/site-packages/girder/utility/install.py", line 294, in main
    parsed.func(parsed)
  File "/root/girder_env/local/lib/python2.7/site-packages/girder/utility/install.py", line 166, in install_web
    plugins=opts.plugins)
  File "/root/girder_env/local/lib/python2.7/site-packages/girder/utility/install.py", line 153, in runWebBuild
    (' '.join(cmd), proc.returncode))
Exception: Web client install failed: `npm run build -- --env=prod --plugins=hashsum_download` returned 1.
```

I can post logs and stuff if you tell me what you'd like to see.

```
npm run build -- --plugins=hashsum_download

> girder@2.0.0 build /root/girder
> grunt init && grunt "--plugins=hashsum_download"

Loading "build.js" tasks...ERROR
>> Error: Cannot find module 'underscore'
Loading "plugin.js" tasks...ERROR
>> Error: Cannot find module 'underscore'
Loading "webpack.config.js" tasks...ERROR
>> Error: Cannot find module 'webpack'
Loading "webpack.dev.js" tasks...ERROR
>> Error: Cannot find module 'webpack'
Loading "webpack.prod.js" tasks...ERROR
>> Error: Cannot find module 'webpack'
>> Local Npm module "grunt-shell" not found. Is it installed?
>> Local Npm module "grunt-contrib-watch" not found. Is it installed?
>> Local Npm module "grunt-contrib-stylus" not found. Is it installed?
>> Local Npm module "grunt-contrib-uglify" not found. Is it installed?
>> Local Npm module "grunt-contrib-copy" not found. Is it installed?
>> Local Npm module "grunt-contrib-symlink" not found. Is it installed?
>> Local Npm module "grunt-gitinfo" not found. Is it installed?
>> Local Npm module "grunt-curl" not found. Is it installed?
>> Local Npm module "grunt-zip" not found. Is it installed?
>> Local Npm module "grunt-file-creator" not found. Is it installed?
>> Local Npm module "grunt-npm-install" not found. Is it installed?
>> Local Npm module "grunt-webpack" not found. Is it installed?
Loading "Gruntfile.js" tasks...ERROR
>> Error: Cannot find module 'toposort'
Warning: Task "init" not found. Use --force to continue.

Aborted due to warnings.

npm ERR! Linux 4.2.0-42-generic
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "run" "build" "--" "--plugins=hashsum_download"
npm ERR! node v6.8.1
npm ERR! npm  v3.10.9
npm ERR! code ELIFECYCLE                                                                                                                                                           
npm ERR! girder@2.0.0 build: `grunt init && grunt "--plugins=hashsum_download"`                                                                                                    
npm ERR! Exit status 3                                                                                                                                                             
npm ERR!                                                                                                                                                                           
npm ERR! Failed at the girder@2.0.0 build script 'grunt init && grunt "--plugins=hashsum_download"'.                                                                               
npm ERR! Make sure you have the latest version of node.js and npm installed.                                                                                                       
npm ERR! If you do, this is most likely a problem with the girder package,                                                                                                         
npm ERR! not with npm itself.                                                                                                                                                      
npm ERR! Tell the author that this fails on your system:                                                                                                                           
npm ERR!     grunt init && grunt "--plugins=hashsum_download"                                                                                                                      
npm ERR! You can get information on how to open an issue for this project with:                                                                                                    
npm ERR!     npm bugs girder                                                                                                                                                       
npm ERR! Or if that isn't available, you can get their info via:                                                                                                                   
npm ERR!     npm owner ls girder                                                                                                                                                   
npm ERR! There is likely additional logging output above.                                                                                                                          
                                                                                                                                                                                   
npm ERR! Please include the following file with any support request:                                                                                                               
npm ERR!     /root/girder/npm-debug.log       
```

the fresh `npm install` gave trouble with `phantom-js`, which I think I also saw during the initial install but sucessfully(?) ignored.

```
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.0.0 (node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.0.14: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN grunt-webpack@1.0.18 requires a peer of webpack@1.x but none was installed.
npm WARN grunt-webpack@1.0.18 requires a peer of webpack-dev-server@1.x but none was installed.
npm ERR! Linux 4.2.0-42-generic
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install"
npm ERR! node v6.8.1
npm ERR! npm  v3.10.9
npm ERR! code ELIFECYCLE

npm ERR! phantomjs-prebuilt@2.1.13 install: `node install.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the phantomjs-prebuilt@2.1.13 install script 'node install.js'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the phantomjs-prebuilt package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node install.js
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs phantomjs-prebuilt
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls phantomjs-prebuilt
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /root/girder/npm-debug.log
```

@zachmullen I tested out the merge you guys made, still getting Uglify errors:

```
ERROR in girder.app.min.js from UglifyJs
Unexpected character '`' [girder.app.min.js:5900,317]

ERROR in plugins/item_previews/plugin.min.js from UglifyJs
Unexpected character '`' [plugins/item_previews/plugin.min.js:142,72]

ERROR in plugins/thumbnails/plugin.min.js from UglifyJs
Unexpected character '`' [plugins/thumbnails/plugin.min.js:2349,110]

ERROR in plugins/jobs/plugin.min.js from UglifyJs
Unexpected character '`' [plugins/jobs/plugin.min.js:244,224]

ERROR in plugins/homepage/plugin.min.js from UglifyJs
Unexpected character '`' [plugins/homepage/plugin.min.js:2188,317]
Warning: Task "webpack:prod" failed. Use --force to continue.

Aborted due to warnings.

npm ERR! Linux 4.2.0-42-generic
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "run" "build" "--" "--env=prod" "--plugins=jobs,item_previews,hashsum_download,gravatar,autojoin,metadata_extractor,homepage,thumbnails"
npm ERR! node v6.8.1
npm ERR! npm  v3.10.9
npm ERR! code ELIFECYCLE
npm ERR! girder@2.0.0 build: `grunt init && grunt "--env=prod" "--plugins=jobs,item_previews,hashsum_download,gravatar,autojoin,metadata_extractor,homepage,thumbnails"`
npm ERR! Exit status 3
npm ERR! 
npm ERR! Failed at the girder@2.0.0 build script 'grunt init && grunt "--env=prod" "--plugins=jobs,item_previews,hashsum_download,gravatar,autojoin,metadata_extractor,homepage,thumbnails"'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the girder package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     grunt init && grunt "--env=prod" "--plugins=jobs,item_previews,hashsum_download,gravatar,autojoin,metadata_extractor,homepage,thumbnails"
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs girder
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls girder
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /root/girder_env/lib/python2.7/site-packages/girder/npm-debug.log
Traceback (most recent call last):
  File "/root/girder_env/bin/girder-install", line 11, in <module>
    load_entry_point('girder', 'console_scripts', 'girder-install')()
  File "/root/girder_env/local/lib/python2.7/site-packages/girder/utility/install.py", line 294, in main
    parsed.func(parsed)
  File "/root/girder_env/local/lib/python2.7/site-packages/girder/utility/install.py", line 166, in install_web
    plugins=opts.plugins)
  File "/root/girder_env/local/lib/python2.7/site-packages/girder/utility/install.py", line 153, in runWebBuild
    (' '.join(cmd), proc.returncode))
Exception: Web client install failed: `npm run build -- --env=prod --plugins=jobs,item_previews,hashsum_download,gravatar,autojoin,metadata_extractor,homepage,thumbnails` returned 1.
```

--- a/docs/build-docs.rst
+++ b/docs/build-docs.rst
@@ -1,17 +1,7 @@
 Build the Sphinx Documentation
 ==============================
 
-In order to build the `Sphinx <http://sphinx-doc.org>`_ documentation, you can
-use the Grunt task at the top level like so: ::
+In order to build the `Sphinx <http://sphinx-doc.org>`_ documentation, after
+setting up your development environment, you can run the following: ::
 
-    grunt docs
-
-or manually run the Makefile here: ::
-
-    make html
-
-This assumes that the Sphinx package is installed in your site packages or
-virtual environment. If that is not yet installed, it can be done using
-**pip**. ::
-
-    pip install sphinx
+    npm run docs

--- a/docs/plugins.rst
+++ b/docs/plugins.rst
@@ -491,3 +491,77 @@ users should now be able to run the task represented by the docker image.
 .. note:: The same item task can be auto-configured via docker multiple times without
    causing problems. Different image ids may be used each time if desired.
 
+
+Hashsum Download
+----------------
+
+The hashum_download plugin allows a file to be downloaded from Girder given a hash value and hash
+algorithm. Use this plugin when you have large data that you don’t want to keep in a software
+repository, but want to access that data from the repository, e.g. during a build or test of that
+software project. This plugin is written to satisfy the needs of CMake ExternalData. These docs
+describe how to use this plugin along with ExternalData, but the plugin could be used outside of
+that context. For more detailed documentation on how to use this in a software repository see the
+`ITKExamples <https://itk.org/ITKExamples/Documentation/Contribute/UploadBinaryData.html>`_. This
+example project uses the Girder instance https://data.kitware.com.
+
+.. note:: The use of the hashsum_download plugin with CMake ExternalData is only supported with a
+   filesystem assetstore and SHA512 as the hash algorithm.
+
+As every local Git repository contains a copy of the entire project history, it is important to
+avoid adding large binary files directly to the repository. Large binary files added and removed
+throughout a project’s history will cause the repository to become bloated and take up too much
+disk space, requiring excessive time and bandwidth to download, etc.
+
+A solution to this problem, when using the CMake build system, is to store binary files in a
+separate location outside the Git repository, then download the files at build time with CMake.
+
+A "content link" file contains an identifying SHA512 hash. The content link is calculated based on
+the original data file, but it should be stored in the Git repository at the path where the file
+would exist, with a ".sha512" extension appended to the file name. CMake will find these content
+link files at build time, download them from a list of server resources, and create symlinks or
+copies of the original files in the build tree.  The data file is stored in Girder, with the SHA512
+for the data added as metadata. Given a content link file containing a SHA512 hash, the data file
+that generated that hash can be retrieved from Girder by its hash.
+
+CMake ExternalData provides tooling to connect with a Girder instance, download the actual data
+file pointed to by the content link file, and provide a file path so tests can access the data
+file contents.
+
+Usage by a software project maintainer
+**************************************
+
+Again, for more background, using the example Girder instance https://data.kitware.com, see the
+`ITKExamples <https://itk.org/ITKExamples/Documentation/Contribute/UploadBinaryData.html>`_. Also
+see the CMake External Data documentation for CMake project configuration
+`help <https://cmake.org/cmake/help/latest/module/ExternalData.html>`_.
+
+In your project, you must set ExternalData_URL_TEMPLATES to a girder url, e.g.
+"https://data.kitware.com:443/api/v1/file/hashsum/%(algo)/%(hash)/download".
+
+See the ITK configuration for an `example <https://github.com/InsightSoftwareConsortium/ITKExamples/blob/master/CMake/ITKExamplesExternalData.cmake>`_
+application of ExternalData_URL_TEMPLATES.
+
+Project contributors will add data files to a Girder instance in arbitrary folders. At a project
+release and on a regular basis, perhaps nightly, the data should be archived in a new Girder folder
+to ensure its persistence.  A `script <https://github.com/InsightSoftwareConsortium/ITK/blob/ef14cce1c26d5dce7eb2e10d36c7dc81aaa9c9e8/Utilities/Maintenance/ArchiveTestingDataOnGirder.py>`_ that provides this functionality is available, as is an
+`example folder <https://data.kitware.com/#collection/57b5c9e58d777f126827f5a1/folder/57b672b48d777f10f269651a>`_
+produced by the script for a release.
+
+Usage by a software project contributor
+***************************************
+
+Upload a file to a Girder instance, which will create a Girder Item to house the file. Navigate to
+the Item, then click on the i (information) icon next to the file, which will show the id, and
+since the hashsum_download plugin is enabled, the sha512 field will also be displayed. Click on the
+key icon to download a hashfile, which will be the full sha512 of the file, with the same name as
+the file, and an extension of .sha512, and you can use this for your CMake content link. E.g.,
+upload my_datafile.txt and download the my_data.txt.sha512 file, then check the my_data.txt.sha512
+file into your source repository.
+
+You can use the Girder API to get the hash of the file given the file id, with the endpoint
+api/v1/file/<file id>/hashsum_file/sha512, where the file id comes from the specific file in Girder.
+
+You can also use the API to download the file based on the hash returned by the previous endpoint,
+with an endpoint /api/v1/file/hashsum/sha512/<file sha512 hash>/download, where the sha512 hash
+comes from the specific file in Girder.
+

--- a/docs/plugins.rst
+++ b/docs/plugins.rst
@@ -510,22 +510,26 @@ example project uses the Girder instance https://data.kitware.com.
 As every local Git repository contains a copy of the entire project history, it is important to
 avoid adding large binary files directly to the repository. Large binary files added and removed
 throughout a project’s history will cause the repository to become bloated and take up too much
-disk space, requiring excessive time and bandwidth to download, etc.
+disk space, requiring excessive time and bandwidth to download.
 
 A solution to this problem, when using the CMake build system, is to store binary files in a
 separate location outside the Git repository, then download the files at build time with CMake.
 
-A "content link" file contains an identifying SHA512 hash. The content link is calculated based on
-the original data file, but it should be stored in the Git repository at the path where the file
-would exist, with a ".sha512" extension appended to the file name. CMake will find these content
-link files at build time, download them from a list of server resources, and create symlinks or
-copies of the original files in the build tree.  The data file is stored in Girder, with the SHA512
-for the data added as metadata. Given a content link file containing a SHA512 hash, the data file
-that generated that hash can be retrieved from Girder by its hash.
+CMake uses the notion of a content link file, which contains an identifying hash calculated from
+the original data file. The content link file has the same name as the data file, with a ".sha512"
+extension appended to the file name, and should be stored in the Git repository. CMake will find
+these content link files at build time, download the corresponding data files from a list of server
+resources, and create symlinks or copies of the original files in the build tree, which is why the
+files are called "content links".
 
-CMake ExternalData provides tooling to connect with a Girder instance, download the actual data
-file pointed to by the content link file, and provide a file path so tests can access the data
-file contents.
+What CMake calls a content link file, Girder calls a key file, as the notion of content link
+doesn't apply in the context of Girder, and the hash value is a key into the original data file.
+When using the hashsum_download plugin, the data file is stored in Girder, with the SHA512 for the
+data added as metadata and provided as the key file, which can be downloaded from Girder and added
+to a project repository. The hashsum_plugin allows the data file to be downloaded based on the hash
+of the data. CMake ExternalData provides tooling to connect with a Girder instance, download the
+actual data file pointed to by the content link (key) file by passing the hash to Girder, and
+provide a local file path to access the data file contents.
 
 Usage by a software project maintainer
 **************************************
@@ -536,7 +540,7 @@ see the CMake External Data documentation for CMake project configuration
 `help <https://cmake.org/cmake/help/latest/module/ExternalData.html>`_.
 
 In your project, you must set ExternalData_URL_TEMPLATES to a girder url, e.g.
-"https://data.kitware.com:443/api/v1/file/hashsum/%(algo)/%(hash)/download".
+"https://data.kitware.com/api/v1/file/hashsum/%(algo)/%(hash)/download".
 
 See the ITK configuration for an `example <https://github.com/InsightSoftwareConsortium/ITKExamples/blob/master/CMake/ITKExamplesExternalData.cmake>`_
 application of ExternalData_URL_TEMPLATES.
@@ -551,17 +555,18 @@ Usage by a software project contributor
 ***************************************
 
 Upload a file to a Girder instance, which will create a Girder Item to house the file. Navigate to
-the Item, then click on the i (information) icon next to the file, which will show the id, and
+the Item, then click on the **i** (information) icon next to the file, which will show the id, and
 since the hashsum_download plugin is enabled, the sha512 field will also be displayed. Click on the
 key icon to download a hashfile, which will be the full sha512 of the file, with the same name as
-the file, and an extension of .sha512, and you can use this for your CMake content link. E.g.,
-upload my_datafile.txt and download the my_data.txt.sha512 file, then check the my_data.txt.sha512
-file into your source repository.
+the file, and an extension of .sha512, and you can use this key file as your CMake content link.
+E.g., upload my_datafile.txt and download the my_data.txt.sha512 file, then check the
+my_data.txt.sha512 file into your source repository.
 
 You can use the Girder API to get the hash of the file given the file id, with the endpoint
-api/v1/file/<file id>/hashsum_file/sha512, where the file id comes from the specific file in Girder.
+``api/v1/file/<file id>/hashsum_file/sha512``, where the file id comes from the specific file in
+Girder.
 
 You can also use the API to download the file based on the hash returned by the previous endpoint,
-with an endpoint /api/v1/file/hashsum/sha512/<file sha512 hash>/download, where the sha512 hash
+with an endpoint ``/api/v1/file/hashsum/sha512/<file sha512 hash>/download``, where the sha512 hash
 comes from the specific file in Girder.
 

--- a/plugins/hashsum_download/plugin.json
+++ b/plugins/hashsum_download/plugin.json
@@ -1,5 +1,0 @@
-{
-"name": "Hashsum download",
-"description": "Allows download of a file by its hashsum.",
-"version": "0.1.0"
-}

--- a/plugins/hashsum_download/plugin.yml
+++ b/plugins/hashsum_download/plugin.yml
@@ -1,0 +1,4 @@
+name: Hashsum download
+description: Allows download of a file by its hashsum.
+url: http://girder.readthedocs.io/en/latest/plugins.html#hashsum-download
+version: 0.1.0

